mantidproject/mantid
# 35373 support for collimator in DiscusMultipleScatteringCorrection algorithm

### Description of work
This work adds support for a radial collimator in `DiscusMultipleScatteringCorrection` algorithm that it restricts scatter points within a small region within the larger sample volume. The algorithm was modified to assign zero weight to tracks where the final scatter is not in a position that allows the final track segment to pass through the collimator toward detectors.


#### Purpose of work
This work allows the use of `DiscusMultipleScatteringCorrection` algorithm for ENGINX instrument that has a radial collimator.

Fixes #35373. 




#### Further detail of work


### To test:
To test this feature, run the `DiscusMultipleScatteringCorrection` with and without `RadialCollimator` and compare the output weights assigned by the algorithm for each spectrum(detector) that are un masked. Note: To test the algorithm in reduced time period, many of the detectors are marked as masked. 

Run below code which will Mask all the detectors from spectrum 10 onwards.
```
from mantid.simpleapi import *

X=[0.99,1.0,1.01]
Y=[0.,100,0.]
Sofq=CreateWorkspace(DataX=X,DataY=Y,UnitX="MomentumTransfer")

run = "ENGINX00194547"
ws=Load(Filename=run+'.nxs', OutputWorkspace=run)
MoveInstrumentComponent(ws, ComponentName="sample position", X=0, Y=0, Z=0, RelativePosition=False)

sphere_xml = " \
<sphere id='some-sphere'> \
    <centre x='0.0'  y='0.0' z='0.0' /> \
    <radius val='0.2' /> \
</sphere> \
<algebra val='some-sphere' /> \
"

SetSample(InputWorkspace=ws,
          Geometry={'Shape': 'CSG', 'Value': sphere_xml},
          Material={'NumberDensity': 0.02, 'AttenuationXSection': 0.0,
                    'CoherentXSection': 0.0, 'IncoherentXSection': 0.0, 'ScatteringXSection': 80.0})

ws_mom = ConvertUnits(InputWorkspace=run, OutputWorkspace=run+'_momentum', Target='Momentum') 
MaskDetectors(ws_mom,  WorkspaceIndexList=range(10, ws_mom.getNumberHistograms()))

results_group1 = DiscusMultipleScatteringCorrection(InputWorkspace=ws_mom, StructureFactorWorkspace=Sofq,
                                                   OutputWorkspace="With_Collimator", NeutronPathsSingle=5,
                                                   NeutronPathsMultiple=10, ImportanceSampling=True, RadialCollimator=True, NumberScatterings=4)

results_group2 = DiscusMultipleScatteringCorrection(InputWorkspace=ws_mom, StructureFactorWorkspace=Sofq,
                                                   OutputWorkspace="No_Collimator", NeutronPathsSingle=5,
                                                   NeutronPathsMultiple=10, ImportanceSampling=True, RadialCollimator=False, NumberScatterings=4)
```

Click `Show Data` for workspaces suffixed `Scatter_n_Integrated` under each group workspaces named `With_Collimator` and `No_Collimator`. When the collimator is present (i.e `RadialCollimator=True`), the integrated weight for `Spectrum index 2` (an unmasked detector) should be lower than the weight when there is no collimator(`RadialCollimator=False`).







---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
